### PR TITLE
Add time support for promethious and refactor

### PR DESCRIPTION
**Pull request by henryoswald**
Child of https://github.com/overleaf/sharelatex/issues/543

### CODE REVIEW

File: timeAsyncMethod.coffee [hunk 1]
```diff
@@ -17,7 +17,7 @@ module.exports = (obj, methodName, prefix, logger) ->
 				logger.log "[Metrics] expected wrapped method '#{methodName}' to be invoked with a callback"
 			return realMethod.apply this, originalArgs
 
-		timer = new metrics.Timer(key)
+		timer = new metrics.Timer(prefix, null, {method: methodName})
```
Comment (mans0954) [inline timeAsyncMethod.coffee]: This appears to set the sample rate to null. Is that a valid value?

File: metrics.coffee [hunk 2]
```diff
@@ -44,63 +40,92 @@ module.exports = Metrics =
...
 				help: key, 
-				labelNames: ['name','host']
+				labelNames: ['app','host','status','method']
 			})
-		counters[key].inc({name: name, host: hostname})
+		opts.app = appname
+		opts.host = hostname
+		promMetrics[key].inc(opts)
+		if process.env['DEBUG_METRICS']
+			console.log("doing inc", key, opts)
```
[the remark this answers is not in the document]
Comment (henryoswald) [reply metrics.coffee]: We don't have the current value with inc

File: metrics.coffee [hunk 3]
```diff
@@ -44,63 +40,92 @@ module.exports = Metrics =
 			res.end(Register.metrics())
 		)
 
+	buildPromKey: (key = "")->
+		Metrics.sanitizeKey key
+
 	sanitizeKey: (key) ->
```
[the remark this answers is not in the document]
Comment (henryoswald) [reply metrics.coffee]: done

File: metrics.coffee [hunk 4]
```diff
@@ -44,63 +40,92 @@ module.exports = Metrics =
...
 		key.replace /[^a-zA-Z0-9]/g, "_"
 
 	sanitizeValue: (value) ->
 		parseFloat(value)
 
 	set : (key, value, sampleRate = 1)->
```
[the remark this answers is not in the document]
Comment (henryoswald) [reply metrics.coffee]: I've added a console.log, I don't think prom offers set like functionality, what do you think? https://github.com/etsy/statsd/blob/master/docs/metric_types.md#sets

File: metrics.coffee [hunk 5]
```diff
@@ -44,63 +40,92 @@ module.exports = Metrics =
...
+		key = Metrics.buildPromKey(key)
+		if !promMetrics[key]?
+			promMetrics[key] = new prom.Counter({
+				name: key,
+				help: key, 
+				labelNames: ['app','host']
+			})
+		promMetrics[key].inc({app: appname, host: hostname}, count)
+		if process.env['DEBUG_METRICS']
+			console.log("doing count/inc", key, opts)
```
[the remark this answers is not in the document]
Comment (henryoswald) [reply metrics.coffee]: ditto to above

**Pull request merged**